src/reconstruction/reconstruction.h

```diff
@@ -241,100 +257,127 @@ Primitive __device__ __host__ __inline__ Van_Leer_Slope(Primitive const &left_sl
+ * \param[in] sound_speed The sound speed
+ * \param[in] sound_speed_squared The sound speed squared
  * \param[in] gamma The adiabatic index
- * \return Characteristic
+ * \return eigenVecs
  */
-Characteristic __device__ __inline__ Primitive_To_Characteristic(Primitive const &primitive,
-                                                                 Primitive const &primitive_slope,
-                                                                 Real const &sound_speed,
-                                                                 Real const &sound_speed_squared, Real const &gamma)
-{
-  Characteristic output;
-
 #ifdef MHD
+eigenVecs __device__ __inline__ Compute_Eigenvectors(Primitive const &primitive, Real const &sound_speed,
+                                                     Real const &sound_speed_squared, Real const &gamma)
+{
+  eigenVecs output;
   // This is taken from Stone et al. 2008, appendix A. Equation numbers will be quoted as relevant
 
-  // First, compute some basic quantities we will need later
-  Real const inverse_sqrt_density = rsqrt(primitive.density);
-
   // Compute wave speeds and their squares
-  Real const magnetosonic_speed_fast = mhd::utils::fastMagnetosonicSpeed(
+  output.magnetosonic_speed_fast = mhd::utils::fastMagnetosonicSpeed(
       primitive.density, primitive.pressure, primitive.magnetic_x, primitive.magnetic_y, primitive.magnetic_z, gamma);
-  Real const magnetosonic_speed_slow = mhd::utils::slowMagnetosonicSpeed(
+  output.magnetosonic_speed_slow = mhd::utils::slowMagnetosonicSpeed(
       primitive.density, primitive.pressure, primitive.magnetic_x, primitive.magnetic_y, primitive.magnetic_z, gamma);
 
-  Real const magnetosonic_speed_fast_squared = magnetosonic_speed_fast * magnetosonic_speed_fast;
-  Real const magnetosonic_speed_slow_squared = magnetosonic_speed_slow * magnetosonic_speed_slow;
+  output.magnetosonic_speed_fast_squared = output.magnetosonic_speed_fast * output.magnetosonic_speed_fast;
+  output.magnetosonic_speed_slow_squared = output.magnetosonic_speed_slow * output.magnetosonic_speed_slow;
 
   // Compute Alphas (equation A16)
-  Real alpha_fast, alpha_slow;
-  if (Real const denom = (magnetosonic_speed_fast_squared - magnetosonic_speed_slow_squared),
-      numerator_2      = (magnetosonic_speed_fast_squared - sound_speed_squared);
+  if (Real const denom = (output.magnetosonic_speed_fast_squared - output.magnetosonic_speed_slow_squared),
+      numerator_2      = (output.magnetosonic_speed_fast_squared - sound_speed_squared);
       denom <= 0.0 or numerator_2 <= 0.0) {
-    alpha_fast = 1.0;
-    alpha_slow = 0.0;
-  } else if (Real const numerator_1 = (sound_speed_squared - magnetosonic_speed_slow_squared); numerator_1 <= 0.0) {
-    alpha_fast = 0.0;
-    alpha_slow = 1.0;
+    output.alpha_fast = 1.0;
+    output.alpha_slow = 0.0;
+  } else if (Real const numerator_1 = (sound_speed_squared - output.magnetosonic_speed_slow_squared);
```

Comment:
Since you're not using `numerator_1` in the scope of the if statement, I wonder if
`if ( (sound_speed_squared - output.magnetosonic_speed_slow_squared) <= 0.0) {` 
might be more readable?

Reply:
`numerator_1` is used in the `else` part of the statement. These declartions are a little weird in that they're defined for the entire if, else if, else block and not confined to a single part of it.